velox/common/memory/HashStringAllocator.cpp

```diff
@@ -408,50 +408,48 @@ HashStringAllocator::allocateFromFreeList(
 
 void HashStringAllocator::free(Header* _header) {
   Header* header = _header;
-  if (header->size() > kMaxAlloc && !pool_.isInCurrentRange(header) &&
-      allocationsFromPool_.find(header) != allocationsFromPool_.end()) {
-    // A large free can either be a rest of block or a standalone allocation.
-    VELOX_CHECK(!header->isContinued());
-    freeToPool(header, header->size() + sizeof(Header));
-    return;
-  }
 
   do {
     Header* continued = nullptr;
     if (header->isContinued()) {
       continued = header->nextContinued();
       header->clearContinued();
     }
-    VELOX_CHECK(!header->isFree());
-    freeBytes_ += header->size() + sizeof(Header);
-    cumulativeBytes_ -= header->size();
-    Header* next = header->next();
-    if (next) {
-      VELOX_CHECK(!next->isPreviousFree());
-      if (next->isFree()) {
-        --numFree_;
-        removeFromFreeList(next);
-        header->setSize(header->size() + next->size() + sizeof(Header));
-        next = reinterpret_cast<Header*>(header->end());
-        VELOX_CHECK(next->isArenaEnd() || !next->isFree());
+    if (header->size() > kMaxAlloc && !pool_.isInCurrentRange(header) &&
```

Comment:
Would this condition `allocationsFromPool_.find(header) != allocationsFromPool_.end()` be sufficient to confirm that its an allocation from the pool? If yes, we can then make the other 2, namely `header->size() > kMaxAlloc` and `!pool_.isInCurrentRange(header)` as velox checks since they should also be true for consistency.

This will ensure that incase we run into any inconsistency where either one of those are false then we dont end up taking the route that frees from the arena (potentially making things worse)

Reply:
Checking allocationsFromPool_ would be sufficient, but this check is slow. Hence, we first run quick checks hoping they are false.